### PR TITLE
PR # 2: Seeding Script PR #1 WIP

### PR DESCRIPTION
**Pull request by hannahmanfredi**
Although seeding script times out after seeding just over 3k product records, and all of my tables haven't been seeded yet, this PR represents a big chunk of work in which I:
-organized my code into MVC Architecture
-designed my SQL schema for Postgres
-set up csv generators for each of my postgres tables so I can speed up bacth seeding of my db

I also ran into some insane bugs while trying to dl couchDb as my secondary db and can only run it on a diff laptop. Submitting this PR so I can acess al this work from my main branch on my other computer.

Thank you!

### CODE REVIEW

**Comment by hannahmanfredi** [comment]
@dev-poston @jametevia @wzhao830 PR up for grabs for whoever gets here first. Thank you!

**Comment by jametevia** [comment]
got it!

File: database/postgres/csv.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+const ws_users  = fs.createWriteStream("users.csv");
+const ws_features  = fs.createWriteStream("features.csv");
+const ws_productFeatures = fs.createWriteStream("productFeatures.csv");
+
+let products = utils.createProducts()
+let reviews = utils.createReviews()
+let users = utils.createUsers()
+let features = utils.createFeatures()
+let productFeatures = utils.createProductFeatures()
+
```
Comment (jametevia) [inline database/postgres/csv.js]: Looks great! One thing to watch out for is that when doing 10 million records you'll have to append to your csv files. I'd checkout out how to append data to csv files and which options you can put in the options object, which currently reads {headers: true} and see how you can append data so you can load your data to the files in chunks

File: database/postgres/seed.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      .on('data', (data) => {
+          csvData.push(data);
+      })
+      .on('data-invalid', (row, rowNumber) =>
+          console.log(
+              `Invalid [rowNumber=${rowNumber}] [row=${JSON.stringify(row)}]`
+          )
+      )
+      .on('end', () => {
+          const query =
```
Comment (jametevia) [inline database/postgres/seed.js:22]: not sure if you're doing this. The COPY method can be used to bulk insert data into postgres, as long as your csv file exactly matches your table schema !

File: database/postgres/utils.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+const fs = require('fs');
+const args = require('minimist')(process.argv.slice(2));
+const fastcsv = require('fast-csv');
+const postgresDb = require('./index');
+const contains = require('validator/lib/contains');
+
+module.exports = {
+  createProducts: () => {
+    let products = [];
+    const min = 100
```
Comment (jametevia) [inline database/postgres/utils.js:11]: Looks good! How long is this taking you to generate data? My data structure is similarly complicated and it's taking me a long time to generate my data!

File: schema.sql [hunk 4]
```diff
@@ -0,0 +1,51 @@
+DROP DATABASE IF EXISTS sdc;
+CREATE DATABASE sdc;
+
+\c sdc
+
+DROP TABLE IF EXISTS Products CASCADE;
+CREATE TABLE IF NOT EXISTS Products (
```
Comment (jametevia) [inline schema.sql:5]: Looks good!

File: server/controllers/index.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
+const models = require('../models/index.js');
```
Comment (jametevia) [inline server/controllers/index.js:1]: this is interesting. I'm not sure I've done much with controller files. Definitely curious to learn more about this sometime

File: server/models/index.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      console.log('inside mongo getAvgRating')
+      return db.getAverageRating(req.params.productid)
+      .then((score) => {
+        res.setHeader('content-type', 'application/json');
+        res.send(JSON.stringify(score));
+      })
+    },
+    createReview: () => {
+      console.log('inside mongo createReview')
+      return db.createReview(req.params.productid, (err, data) => {
```
Comment (jametevia) [inline server/models/index.js:72]: again, super interesting. I need to learn more about MVC !

File: server/routes.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+var controller = require('./controllers/index.js');
+var router = require('express').Router();
```
Comment (jametevia) [inline server/routes.js:2]: cleanly formatted. nice!

**Review by jametevia — COMMENTED** [review]
Overall, great work on everything :) Only thing I would say to checkout if you're having trouble with insertion errors is to checkout postgres's copy method for importing csv file data into postgres.

**Pull request merged**